distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Date_Range.enso

```diff
@@ -0,0 +1,484 @@
+       original range, but in the opposite order.
+
+       > Example
+         Reverse a range of dates.
+
+             (Date.new 2020 10 01).up_to (Date.new 2020 10 31) . reverse
+
+       ? Returning a `Vector`
+
+         This method cannot return back a `Date_Range`, as some ranges are not
+         reversible. For example, the range `(Date.new 2020 02 29).up_to (Date.new 2023) . with_step Date_Period.Year`
+         will have `2022-02-28` as its last entry. But if we create a
+         range starting at `2022-02-28` and going backwards by a year, its last
+         element will be `2020-02-28` and not `2020-02-29` as in the original.
+         Thus, to preserve the contents we need to return a vector.
+    reverse : Vector Date
+    reverse self = self.to_vector.reverse
+
+    ## Combines all the elements of a non-empty range using a binary operation.
+       If the range is empty, returns `if_empty`.
+
+       Arguments:
+       - function: A binary operation that takes two dates and combines them
+         into a new date.
+       - if_empty: Value returned if the range is empty.
+    reduce : (Date -> Date -> Date) -> Any -> Any
+    reduce self function ~if_empty=(Error.throw Empty_Error) =
+        case self.length of
+            0 -> if_empty
+            1 -> self.start
+            length ->
+                (1.up_to length).fold self.start acc-> ix->
+                    function acc (self.internal_at ix)
+
+
+## PRIVATE
+   Computes the length of the range and verifies its invariants.
+
+   If any of the invariants are violated, a dataflow error is raised.
+compute_length_and_verify : Date -> Date -> Period -> Boolean -> Integer ! Illegal_Argument
+compute_length_and_verify start end step increasing =
+    if is_period_positive step . not then Error.throw (Illegal_Argument.Error "The step `Period` for `Date_Range` must be positive, i.e. all of `years`, `months` and `days` must be non-negative and at least one of them must be strictly positive.") else
+        is_range_empty = case increasing of
+            True -> start >= end
+            False -> start <= end
+        if is_range_empty then 0 else
+            # First a few heuristics for a fast path.
+            # If there are no years or months, we can perform a simple computation on day difference.
+            if step.total_months == 0 then compute_length_step_days start end step.days increasing else
+                # Similarly, if we are only shifting by months, we can rely on a simpler computation.
+                if step.days == 0 then compute_length_step_months start end step.total_months increasing else
+                    # Then we go brute force for the general case.
+                    compute_length_step_brute_force start end step increasing
+
+## PRIVATE
+is_period_positive period =
+    if (period.years < 0) || (period.months < 0) || (period.days < 0) then False else
+        (period.total_months > 0) || (period.days > 0)
```

Comment:
don't think we need to but could allow `(period.total_months < 0) || (period.days < 0)` as the month test.

Reply:
Yes, I was considering this, but I thought that it is easier for the user to define the 'positive' Period as _all_ properties being non-negative - and I wanted the check to be consistent with it.

If I changed it, the period `Period.new years=1 months=-3` would still be deemed valid as its `total_months` is `9`. And such a period _would work fine_ with our logic. But I thought that  excluding it makes the documentation and expected behaviour simper without losing functionality - the user can always just do `Period.new months=9`.

Reply:
fair. Happy with that.